oauth2/src/main/java/com/linecorp/armeria/common/auth/oauth2/SerialFuture.java

```diff
@@ -0,0 +1,127 @@
+
+import com.linecorp.armeria.common.util.Exceptions;
+
+/**
+ * Executes submitted {@link Callable} actions asynchronously in sequence.
+ */
+public class SerialFuture {
+
+    private final Queue<Runnable> actions = new LinkedList<>();
+
+    @Nullable
+    private Runnable active;
+
+    @Nullable
+    Executor executor;
+
+    /**
+     * Constructs {@link SerialFuture} with a supplied {@link Executor}.
+     * @param executor An {@link Executor} to execute asynchronous actions.
+     */
+    public SerialFuture(@Nullable Executor executor) {
+        this.executor = executor;
+    }
+
+    /**
+     * Constructs {@link SerialFuture}.
+     */
+    public SerialFuture() {
+        executor = null;
+    }
+
+    /**
+     * Executes submitted {@link Callable} action asynchronously in sequence.
+     * @param action An asynchronous {@link Callable} action to be executed in sequence.
+     *               It returns {@link CompletionStage} that produces the final result when the action
+     *               completes.
+     * @return Returns a new {@link CompletionStage} with the same result or exception as the {@code action},
+     *         that executes the given action when this stage completes.
+     */
+    public synchronized <V> CompletionStage<V> executeAsync(Callable<CompletionStage<V>> action) {
+        final CompletableFuture<V> result = new CompletableFuture<>();
+        actions.add(() -> {
+            final CompletionStage<V> future;
+            try {
+                future = action.call();
+            } catch (Exception e) {
+                result.completeExceptionally(Exceptions.peel(e));
+                executeNext();
+                return;
+            }
+            future.whenComplete((v, ex) -> {
+                if (ex == null) {
+                    result.complete(v);
+                } else {
+                    result.completeExceptionally(ex);
+                }
+                executeNext();
+            });
+        });
+
+        if (active == null) {
+            executeNext();
+        }
+        return result;
+    }
+
+    /**
+     * Calls submitted {@link Callable} action asynchronously in sequence.
+     * @param action A {@link Callable} action to be called asynchronously in sequence.
+     *               It returns that produces the final result of computation.
+     * @return Returns a new {@link CompletionStage} with the same result or exception as the {@code action},
+     *         that executes the given action when this stage completes.
+     */
+    public synchronized <V> CompletionStage<V> callAsync(Callable<V> action) {
+        final CompletableFuture<V> result = new CompletableFuture<>();
+        actions.add(() -> {
+            try {
+                result.complete(action.call());
+                executeNext();
+            } catch (Exception e) {
+                result.completeExceptionally(Exceptions.peel(e));
+                executeNext();
+            }
```

Comment:
Could catch `Throwable` and call `executeNext()` in `finally`.

Reply:
fixed